src/main/java/net/neoforged/neoforge/registries/NeoForgeRegistriesSetup.java

```diff
@@ -0,0 +1,91 @@
+
+@ApiStatus.Internal
+public class NeoForgeRegistriesSetup {
+    private static boolean setup = false;
+
+    /**
+     * Internal forge method. Modders do not call.
+     */
+    public static synchronized void setup(IEventBus modEventBus) {
+        if (setup)
+            throw new IllegalStateException("Setup has already been called!");
+
+        setup = true;
+
+        modEventBus.addListener(NeoForgeRegistriesSetup::registerRegistries);
+        modEventBus.addListener(NeoForgeRegistriesSetup::onModifyRegistry);
+    }
+
+    /**
+     * The set of vanilla registries which should be synced to the client.
+     */
+    private static final Set<ResourceKey<? extends Registry<?>>> VANILLA_SYNC_KEYS = Set.of(
+            Registries.SOUND_EVENT, // Required for SoundEvent packets
+            Registries.MOB_EFFECT, // Required for MobEffect packets
+            Registries.BLOCK, // Required for chunk BlockState paletted containers syncing
+            Registries.ENCHANTMENT, // Required for EnchantmentMenu syncing
+            Registries.ENTITY_TYPE, // Required for Entity spawn packets
+            Registries.ITEM, // Required for Item/ItemStack packets
+            Registries.PARTICLE_TYPE, // Required for ParticleType packets
+            Registries.BLOCK_ENTITY_TYPE, // Required for BlockEntity packets
+            Registries.PAINTING_VARIANT, // Required for EntityDataSerializers
+            Registries.MENU, // Required for ClientboundOpenScreenPacket
+            Registries.COMMAND_ARGUMENT_TYPE, // Required for ClientboundCommandsPacket
+            Registries.STAT_TYPE, // Required for ClientboundAwardStatsPacket
+            Registries.VILLAGER_TYPE, // Required for EntityDataSerializers
+            Registries.VILLAGER_PROFESSION, // Required for EntityDataSerializers
+            Registries.CAT_VARIANT, // Required for EntityDataSerializers
+            Registries.FROG_VARIANT // Required for EntityDataSerializers
+    );
+
+    private static void registerRegistries(NewRegistryEvent event) {
+        event.register(NeoForgeRegistries.ENTITY_DATA_SERIALIZERS);
+        event.register(NeoForgeRegistries.GLOBAL_LOOT_MODIFIER_SERIALIZERS);
+        event.register(NeoForgeRegistries.BIOME_MODIFIER_SERIALIZERS);
+        event.register(NeoForgeRegistries.STRUCTURE_MODIFIER_SERIALIZERS);
+        event.register(NeoForgeRegistries.FLUID_TYPES);
+        event.register(NeoForgeRegistries.HOLDER_SET_TYPES);
+        event.register(NeoForgeRegistries.DISPLAY_CONTEXTS);
+        event.register(NeoForgeRegistries.INGREDIENT_TYPES);
+        event.register(NeoForgeRegistries.CONDITION_SERIALIZERS);
+        event.register(NeoForgeRegistries.ITEM_PREDICATE_SERIALIZERS);
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    private static void onModifyRegistry(ModifyRegistryEvent event) {
+        if (!event.isBuiltin() || !(event.getRegistry() instanceof BaseMappedRegistry<?> forgeRegistry))
```

Comment:
When is it possible for a registry to not be a subclass of `BaseMappedRegistry`? Given that the only two concrete subclasses of `Registry` are also subclasses of `BaseMappedRegistry`?

Do we anticipate other mods making their own `Registry` subclasses which aren't subclasses of `BaseMappedRegistry` (directly, or indirectly via `MappedRegistry` or `DefaultedMappedRegistry`)? If we don't, we should at least emit a warning log that a registry wasn't encountered, if not throw an exception.

Reply:
changed to a "hard" cast